fileserver/routes.py

```diff
@@ -42,6 +56,101 @@ def generate_file_id(data):
         blake2b(data, digest_size=33, salt=b"SessionFileSvr\0\0").digest()
     ).decode()
 
+def abort_with_reason(code, msg, warn=True):
+    if warn:
+        app.logger.warning(msg)
+    else:
+        app.logger.debug(msg)
+    abort(Response(msg, status=code, mimetype='text/plain'))
+
+def valid_blinded_version_id_for_auth(request, required):
+    """
+    Check if a request is correctly authenticated, if the auth headers are missing and auth isn't
+    required then just return 'None'.
+    """
+    pk, ts_str, sig_in = (
+        request.headers.get(f"X-FS-{h}") for h in ('Pubkey', 'Timestamp', 'Signature')
+    )
+    missing = sum(x is None or x == '' for x in (pk, ts_str, sig_in))
+
+    # If we were missing one of the auth headers and don't require auth then just return a 'None'
+    if missing > 0 and not required:
+        return None
```

Comment:
The case where `0 < missing < 3` case is a bit weird: the client is passing *some* of the auth headers but not all.  I think we should probably raise an error in such a case:
```suggestion
    if missing:
        if required or missing < 3:
            abort_with_reason(
                http.BAD_REQUEST, "Invalid authentication: one or more X-FS-* auth headers is missing")
        else:
            return None
```

Reply:
Yep fair enough - updated